src/main/java/com/eppo/sdk/helpers/bandit/FalconBanditModel.java

```diff
@@ -0,0 +1,92 @@
+      Map.Entry::getKey, e -> {
+        String actionName = e.getKey();
+        EppoAttributes actionAttributes = e.getValue();
+        double actionScore = 0.0;
+
+        // get all coefficients known to the model for this action
+        BanditCoefficients banditCoefficients = parameters.getModelData().getCoefficients().get(actionName);
+
+        if (banditCoefficients == null) {
+          // Unknown action; return default score of 0
+          return actionScore;
+        }
+
+        actionScore += banditCoefficients.getIntercept();
+
+        actionScore += scoreContextForCoefficients(actionAttributes, banditCoefficients.getActionNumericCoefficients());
+        actionScore += scoreContextForCoefficients(actionAttributes, banditCoefficients.getActionCategoricalCoefficients());
+        actionScore += scoreContextForCoefficients(subjectAttributes, banditCoefficients.getSubjectNumericCoefficients());
+        actionScore += scoreContextForCoefficients(subjectAttributes, banditCoefficients.getSubjectCategoricalCoefficients());
+
+        return actionScore;
+      })
+    );
+
+    // Convert to scores to weights (probabilities between 0 and 1 that collectively add up to 1.0)
+    Map<String, Double> actionWeights = computeActionWeights(actionScores);
+    return actionWeights;
+  }
+
+  private static double scoreContextForCoefficients(EppoAttributes context, Map<String, ? extends AttributeCoefficients> coefficients) {
+
+    double totalScore = 0.0;
+
+    for (AttributeCoefficients attributeCoefficients : coefficients.values()) {
+      EppoValue contextValue = context.get(attributeCoefficients.getAttributeKey());
+      double attributeScore = attributeCoefficients.scoreForAttributeValue(contextValue);
+      totalScore += attributeScore;
+    }
+
+    return totalScore;
+  }
+
+  private static Map<String, Double> computeActionWeights(Map<String, Double> actionScores) {
+    Double highestScore = null;
+    String highestScoredAction = null;
+    for (Map.Entry<String, Double> actionScore : actionScores.entrySet()) {
+      if (highestScore == null || actionScore.getValue() > highestScore) {
+        highestScore = actionScore.getValue();
+        highestScoredAction = actionScore.getKey();
+      }
+    }
+
+    // Weigh all the actions using their score
+    Map<String, Double> actionWeights = new HashMap<>();
+    double gamma = 1.0; // hard-coded for now
+    double totalNonHighestWeight = 0.0;
+    for (Map.Entry<String, Double> actionScore : actionScores.entrySet()) {
+      if (actionScore.getKey().equals(highestScoredAction)) {
+        // The highest scored action is weighed at the end
+        continue;
+      }
+
+      // Compute weight and round to four decimal places
+      double  unroundedProbability = 1 / (actionScores.size() + (gamma * (highestScore - actionScore.getValue())));
+      double roundedProbability = Math.round(unroundedProbability * 10000d) / 10000d;
```

Comment:
Are we OK with actions being taken with exactly 0 probability (and not 0.01%)? We would stop collecting training data for them. In theory, this breaks the regret bounds for the bandit algorithm. In practice, if we're collecting that little data, it might not make a difference (the model will essentially stop updating anyway).

Reply:
I like the idea of a floor probability. I'm guessing this would be another top-level parameter like `gamma` and `defaultActionScore`.

I would need to think how this factors into the FALCON algorithm. The easiest thing to do may be to chip away at the highest probability to make room for the floor, but this could erode it. Maybe shave a little off all non-floor actions?

This may be over-engineering for our first version, though.

Reply:
I also like a configurable floor probability. Can we do something as simple as `max(floor, unroundedProbability)` to begin with? It isn't super robust, in that a too-high floor will take away a lot of probability from the highest action (potentially even making its probability negative). But for reasonable choices of the floor it won't be a problem. I'm OK both doing something fancy / over-engineered, or worrying about this later on.

Reply:
I'll add it so we have it to play with, but I will default to just taking it off the highest-scored probability as that's the easiest. But there are scenarios (e.g., _lots_ of low-scoring actions ) where things could get weird so I imagine we'll want to proceed with nonzero probabilities with caution.

Reply:
That's the correct thing to do according to the algorithm anyway: it says the highest scoring probability should be 1-sum of all the other probabilities